### DISCUSSION TITLE
Dependency Management

### DISCUSSION THREAD

**Discussion by sbwalker** [oqtane/oqtane.framework#3564] [General]
One of the huge benefits of a modular framework is the loosely coupled architecture which provides superior flexibility in development, maintenance, and deployment. However, every architectural approach also contains some challenges... and the primary challenge for loose coupling is related to dependency management. 

From a composability perspective Oqtane already contains all of the essential services for ensuring the framework is able to dynamically discover, load, and upgrade its dependencies based on the run-time configuration of the framework. However one of the areas which had not yet been fully implemented was the uninstall of extensions. 

Oqtane has supported uninstall as a feature since version 1.0 however it took a very simplistic/optimistic approach. Essentially it removed all of the assets which were identified by a module's installation manifest. This worked fine in most cases. However if a number of different modules were using the same assets (ie. they were shared) it could cause issues because uninstalling one module could remove dependencies which another module required. 

In Oqtane 5.0.1 a simplistic dependency manager has been added to attempt to mitigate this scenario:

https://github.com/oqtane/oqtane.framework/pull/3557

There are 3 aspects to this feature:

1. There is a scheduled job named the "Purge Jog" which runs once per day in each Oqtane installation. It is responsible for general housekeeping to ensure data is optimized and archived appropriately. Some new logic was added to maintain an assembly log of all assemblies in the /bin folder (ie. all core framework assemblies as well as assemblies used by the extensions). On a daily basis, new assemblies will be added to the log and missing assemblies are removed.

2. When installing extensions as Nuget packages (either interactively or by copying them into the Packages folder) the framework will update the assembly log and keep track of the number of extensions which have a dependency on an assembly.

3. When uninstalling an extension, the framework will utilize the assembly log to determine if an assembly is shared (has multiple extensions which depend on it) and will be able to determine if an assembly can be safely deleted.

This approach should help deal with the uninstall scenario. However it is obviously not a perfect solution. If assemblies are being manually deployed or removed from environments, this solution will not be aware of those changes. It is also limited to assemblies currently and does not handle other types of assets such as CSS or JS libraries.

**Comment by leigh-pointer** [comment]
@sbwalker can this be tested now?

**Reply by sbwalker** [reply]
@leigh-pointer yes the code is in the Dev branch. However, keep in mind that in order to test it, the Purge job needs to run at least once to generate the initial assemblies.log file (in the Packages folder). And modules must be installed as Nuget packages (and if the module version being installed was already installed previously, it will be ignored). So there are many aspects to this work flow. ie. it is designed for a typical production environment work flow - not a development environment which is constantly thrashing.

**Reply by leigh-pointer** [reply]
@sbwalker I would be testing on a staging site locally with a personal build of Oqtane 5.0.1

**Reply by sbwalker** [reply]
@leigh-pointer sounds good. You can easily trigger the scheduled job by setting the Next Execution Date to a date in the past

**Reply by leigh-pointer** [reply]
@sbwalker triggered and log created
I see that version information is not captured.  Would it not be a good idea to capture the version.  This info would be valuable when trying to solve dependancy issues?

![image](https://github.com/oqtane/oqtane.framework/assets/9447612/03ac2cee-c3e6-495d-8231-71faa2f06ec1)

**Reply by sbwalker** [reply]
@leigh-pointer Oqtane already contains logic during installation to check the assembly version - and it will only install an assembly to the /bin if it is newer than the assembly which is already installed. #3557 is specifically focused on uninstall ie. ensuring the framework does not delete assemblies which multiple extensions rely upon. In this scenario the version of the assembly is not relevant.

**Comment by leigh-pointer** [comment]
@sbwalker what would be best practice with the Oqtane.License component, as in, should it be deployed with the external module or not?  As the component already resides in the framework, I opted for 'dont deploy' but only reference in my solution.

**Comment by sbwalker** [comment]
@leigh-pointer this is a good question... and unfortunately there is still no perfect answer. 

In version 5.0 we included the licensing component in the framework distribution to try to resolve the issue you raised where uninstalling a commercial module/theme would also uninstall the Licensing component (which could cause issues if other modules/themes were using it). It also prevented the removal of the Licensing assemblies in the uninstall logic. 

However, there was a still a major limitation with this solution - if the commercial module/theme was intended to be compatible with older versions of Oqtane (ie. 4.x) then these improvements added in 5.0.0 would not be applicable, as you would need to distribute the Licensing component with your module to ensure it is installed. 

Option 1 - include the Licensing component in your package. This will allow your module/theme to function properly when installed on Oqtane 4.x or 5.x. On 5.x the new dependency manager will recognize when the Licensing component is installed and will increase the count - which will prevent it from being uninstalled. However on 4.x there is no safeguard or dependency manager - so uninstalling a module will remove the Licensing component.

Option 2 - do not include the Licensing component in your package. This will cause your module/theme to fail when installed on Oqtane 4.x (since it is only 5.x where it is included by default). On 4.x and 5.x uninstall will work fine, as it will not remove the Licensing component if it was not included in the original package. 

Basically there is no way to fix the legacy uninstall behavior of older Oqtane versions (ie. 4.x).

One other scenario which just came to mind is if someone is using Oqtane 4.x and they are using a commercial module. Then they upgrade to Oqtane 5.x... the dependency manager is not going to know that they were using a commercial module. It will include the Licensing component in the dependency manager with a count of 1. And if you then decide to uninstall the module/theme it will use the asset manifest which includes references to the Licensing component, and it will delete the Licensing component (because the count is 1). To deal with this scenario, I may need to add back the hard-coded logic which prevents the deletion of the Licensing component (which I removed as I expected the new dependency manager could handle it).

One other improvement which could be added during installation is that the Licensing component could be ignored during installation if the framework is running version 5.0.1 or higher - as there is no need to install it since it is bundled with the framework. I was trying to avoid hard-coded logic such as this but I am not sure if there are any alternatives.

As I mentioned previously... dependency management is a really painful area to deal with. Going forward, as Oqtane 5.x becomes the mainstream platform, the new capabilities added in 5.0.1 should help a lot in ensuring the install/uninstall supports shared dependencies.

**Comment by leigh-pointer** [comment]
@sbwalker what if the Licensing component was installed as a component and registered in with the modules or rather than a component build the control directly into the framework?  Thinking out loud <g>

**Reply by sbwalker** [reply]
@leigh-pointer those suggestions will not address the core problem related to legacy versions of Oqtane (ie. 4.x)
